### ISSUE TITLE
UnicodeDecodeError when displaying Pandas DataFrames in notebook

### ISSUE THREAD

**Issue by fonnesbeck** [ipython/ipython#6799]
I'm running into issues with displaying a Pandas DataFrame that seems to have some characters that can't be dealt with in IPython notebook. The file imports fine, and displays fine in the console, but on the notebook, I get a `UnicodeDecodeError` when I try to look at it.

    ---------------------------------------------------------------------------
    UnicodeDecodeError                        Traceback (most recent call last)
    <ipython-input-31-a83dda38f1a4> in <module>()
    ----> 1 pd.read_csv('data/Study Characteristics.csv')

    /usr/local/lib/python2.7/site-packages/IPython/core/displayhook.pyc in __call__(self, result)
        242             self.update_user_ns(result)
        243             self.log_output(format_dict)
    --> 244             self.finish_displayhook()
        245 
        246     def flush(self):

    /usr/local/lib/python2.7/site-packages/IPython/kernel/zmq/displayhook.pyc in finish_displayhook(self)
         66         sys.stdout.flush()
         67         sys.stderr.flush()
    ---> 68         self.session.send(self.pub_socket, self.msg, ident=self.topic)
         69         self.msg = None
         70 

    /usr/local/lib/python2.7/site-packages/IPython/kernel/zmq/session.pyc in send(self, stream, msg_or_type, content, parent, ident, buffers, track, header, metadata)
        647         if self.adapt_version:
        648             msg = adapt(msg, self.adapt_version)
    --> 649         to_send = self.serialize(msg, ident)
        650         to_send.extend(buffers)
        651         longest = max([ len(s) for s in to_send ])

    /usr/local/lib/python2.7/site-packages/IPython/kernel/zmq/session.pyc in serialize(self, msg, ident)
        551             content = self.none
        552         elif isinstance(content, dict):
    --> 553             content = self.pack(content)
        554         elif isinstance(content, bytes):
        555             # content is already packed, as in a relayed message

    /usr/local/lib/python2.7/site-packages/IPython/kernel/zmq/session.pyc in <lambda>(obj)
         83 # disallow nan, because it's not actually valid JSON
         84 json_packer = lambda obj: jsonapi.dumps(obj, default=date_default,
    ---> 85     ensure_ascii=False, allow_nan=False,
         86 )
         87 json_unpacker = lambda s: jsonapi.loads(s)

    /usr/local/lib/python2.7/site-packages/zmq/utils/jsonapi.pyc in dumps(o, **kwargs)
         38         kwargs['separators'] = (',', ':')
         39 
    ---> 40     s = jsonmod.dumps(o, **kwargs)
         41 
         42     if isinstance(s, unicode):

    /usr/local/Cellar/python/2.7.6_1/Frameworks/Python.framework/Versions/2.7/lib/python2.7/json/__init__.pyc in dumps(obj, skipkeys, ensure_ascii, check_circular, allow_nan, cls, indent, separators, encoding, default, sort_keys, **kw)
        248         check_circular=check_circular, allow_nan=allow_nan, indent=indent,
        249         separators=separators, encoding=encoding, default=default,
    --> 250         sort_keys=sort_keys, **kw).encode(obj)
        251 
        252 

    /usr/local/Cellar/python/2.7.6_1/Frameworks/Python.framework/Versions/2.7/lib/python2.7/json/encoder.pyc in encode(self, o)
        208         if not isinstance(chunks, (list, tuple)):
        209             chunks = list(chunks)
    --> 210         return ''.join(chunks)
        211 
        212     def iterencode(self, o, _one_shot=False):

    UnicodeDecodeError: 'ascii' codec can't decode byte 0xc2 in position 19900: ordinal not in range(128)

I've tried using encoder options on import, but nothing seems to work. Running Pandas '0.15.0-20-g2737f5a' on OS X 10.10 and IPython '3.0.0-dev' built from master.

[Here is the data that causes the issue](http://d.pr/f/12i3G) incase it helps.

**Comment by ellisonbg** [comment]
I am routinely seeing this problem with users now. I can verify that IPython's `getdefaultencoding` is giving `UTF-8` while Python's `sys.getdefaultencoding` is giving `ascii` (Python 2.7.9). I think we need to explore a fix in IPython. Unicode master @takluyver any thoughts on this? It would be great to have a quick fix that users can apply immediately as well as a long term solution.

**Comment by ellisonbg** [comment]
Pinging @JeremyMMulcahey who is seeing this as well.

**Comment by takluyver** [comment]
I doubt this is something for us to fix. It seems more likely that it's a bug in how pandas is computing one of its reprs. If there are non ASCII characters, we should be getting a Unicode string, not bytes. Though we might need to check the actual `__repr__`, because I think python 2 insists that that return bytes.

**Comment by xflr6** [comment]
Similar problem here: Displaying a DataFrame with non-ascii characters raises ``UnicodeDecodeError`` on IPython notebook after upgrade to 3.0 (ran fine on previous versions), e.g.:
```
import pandas

pandas.DataFrame({'spam': [u'sp\xe4m']})
```

**Comment by ellisonbg** [comment]
I do think we need to look into this - my notebook that is now running into
this issue worked fine as recently as early December.

On Fri, Mar 6, 2015 at 9:17 AM, Thomas Kluyver <notifications@github.com>
wrote:

> I doubt this is something for us to fix. It seems more likely that it's a
> bug in how pandas is computing one of its reprs. If there are non ASCII
> characters, we should be getting a Unicode string, not bytes. Though we
> might need to check the actual __repr__, because I think python 2 insists
> that that return bytes.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ipython/ipython/issues/6799#issuecomment-77597033>.
>



-- 
Brian E. Granger
Cal Poly State University, San Luis Obispo
@ellisonbg on Twitter and GitHub
bgranger@calpoly.edu and ellisonbg@gmail.com

**Comment by minrk** [comment]
I think this is fixed by #7853

**Comment by tbittner** [comment]
I'm hitting the exact `UnicodeDecodeError` as @fonnesbeck and @xflr6:
```
import pandas as pd
test_df = pd.DataFrame([u'Hello \u2013 World'])
test_df
```

**Comment by takluyver** [comment]
I just tried the sample from @tbittner's comment in master, and it works fine, so I assume this has been fixed by #7853. The fix will be in 3.1.

**Comment by ajschumacher** [comment]
Works in 3.1.0! Thank you! :cake:

**Comment by koorukuroo** [comment]
Yeah~! I solved it with upgrade version 3.1.0.
